### PR TITLE
[improvement] Make dependency sourcing more robust

### PR DESCRIPTION
**Pull request by dansanduleac**
# Before this PR
We collect the dependencies that we are interested in locking from each project via a `subprojectUnifiedClasspath` configuration.
However, the way we collect these is by making `subprojectUnifiedClasspath` just extend the other configurations we're interested in (`compileClasspath`, `runtimeClasspath`). 

This has a few problems:
* an exclude in one configuration propagates into `subprojectUnifiedClasspath`, potentially concealing a real dependency (that only appears in runtime) from the lock file
* since we copy and flatten `subprojectUnifiedClasspath`, we can't tell what configuration each dependency came from anymore. So for instance, if we were to also lock `testCompileClasspath` in this way, we couldn't trace if a dependency only came from test, or is also used in production.

Additionally, we find and copy every inter-project dependency declared by the user, and make an unsafe assumption that if the dependency didn't have a target configuration, then it must be the `default` configuration.

## After this PR
==COMMIT_MSG==
When picking up project dependencies, separate them into configurations based on their source (compile, runtime, + test etc in the future), such that gradle resolves them separately. This makes it less likely that additional rules (e.g. excludes) leak from one configuration into another and poison the final lock state for that project.
==COMMIT_MSG==

We also remove the need to copy the user's inter-project dependencies, by instead relying on [attribute-based matching](https://docs.gradle.org/current/userguide/dependency_management_attribute_based_matching.html) to control what configuration you get when encountering an inter-project dependency without a target configuration.

## Possible downsides?
* relying on more advanced gradle magic - [attribute-based matching](https://docs.gradle.org/current/userguide/dependency_management_attribute_based_matching.html), although this has been introduced for a while now and a core part of dependency resolution in java projects - this is how gradle knows whether to use the compile or runtime variant of a dependency & its transitives

### CODE REVIEW

File: src/main/java/com/palantir/gradle/versions/VersionsLockPlugin.java [hunk 1]
```diff
@@ -201,36 +248,67 @@ private static void sourceDependenciesFromProject(
...
-        // dependencies they have.
-        project.getPluginManager().withPlugin("base", plugin -> {
-            subprojectUnifiedClasspath.configure(conf -> {
-                conf.extendsFrom(project.getConfigurations().getByName(Dependency.DEFAULT_CONFIGURATION));
-            });
+        project.getConfigurations().register(SUBPROJECT_UNIFIED_CONFIGURATION_NAME, conf -> {
+            conf.setVisible(false).setCanBeResolved(false);
+
+            // Mark it so it doesn't receive constraints from VersionsPropsPlugin
+            conf.getAttributes().attribute(VersionsPropsPlugin.CONFIGURATION_EXCLUDE_ATTRIBUTE, true);
```
Comment (ferozco) [inline src/main/java/com/palantir/gradle/versions/VersionsLockPlugin.java:254]: I think just using configuration names to determine what configurations to exclude might be more straightforward/readable. Perhaps as a FLUP

File: src/main/java/com/palantir/gradle/versions/VersionsLockPlugin.java [hunk 2]
```diff
@@ -201,36 +248,67 @@ private static void sourceDependenciesFromProject(
...
+
+            // Mark it as a GCV_SOURCE, so that
+            // 1. it can be depended on when resolving `unifiedClasspath`, which requests that the usage is GCV_SOURCE
```
Comment (ferozco) [inline src/main/java/com/palantir/gradle/versions/VersionsLockPlugin.java]: Is it true that the attribute needs to be set? I think that the second point is really why we want to set the attribute

File: src/main/java/com/palantir/gradle/versions/VersionsLockPlugin.java [hunk 3]
```diff
@@ -201,36 +248,67 @@ private static void sourceDependenciesFromProject(
...
+                .add(createConfigurationDependency(project, SUBPROJECT_UNIFIED_CONFIGURATION_NAME));
+
         project.getPluginManager().withPlugin("java", plugin -> {
-            project.getConfigurations().named(SUBPROJECT_UNIFIED_CONFIGURATION_NAME).configure(conf -> {
-                Stream.of(
-                        JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME,
-                        JavaPlugin.RUNTIME_CLASSPATH_CONFIGURATION_NAME)
-                        .map(project.getConfigurations()::getByName)
-                        .forEach(conf::extendsFrom);
+            String compileClasspathForLock = "consistentVersionsCompile";
```
Comment (ferozco) [inline src/main/java/com/palantir/gradle/versions/VersionsLockPlugin.java]: variableName == to string?

File: src/main/java/com/palantir/gradle/versions/VersionsLockPlugin.java [hunk 4]
```diff
@@ -318,9 +419,25 @@ private void recursivelyCopyProjectDependencies(
...
+                    }
+
+                    log.info(
+                            "Found legacy project dependency (with target configuration): {} -> {}",
+                            dependencySet,
+                            formatProjectDependency(projectDependency));
+
                     if (copiedConfigurationsCache.containsKey(targetConf)) {
                         String copiedConf = copiedConfigurationsCache.get(targetConf);
                         log.debug("Re-using already copied target configuration for dep {} -> {}: {}",
```
Comment (ferozco) [inline src/main/java/com/palantir/gradle/versions/VersionsLockPlugin.java:426]: line 449: mark copied configuration as GCV_INTERNAL

Comment (ferozco) [reply src/main/java/com/palantir/gradle/versions/VersionsLockPlugin.java:426]: +1 for test too

File: src/test/groovy/com/palantir/gradle/versions/VersionsLockPluginIntegrationSpec.groovy [hunk 5]
```diff
@@ -51,6 +51,7 @@ class VersionsLockPluginIntegrationSpec extends IntegrationSpec {
                 repositories {
                     maven { url "file:///${mavenRepo.getAbsolutePath()}" }
                 }
+                group = "test"
```
Comment (ferozco) [inline src/test/groovy/com/palantir/gradle/versions/VersionsLockPluginIntegrationSpec.groovy]: do we need this?

File: src/test/groovy/com/palantir/gradle/versions/VersionsLockPluginIntegrationSpec.groovy [hunk 6]
```diff
@@ -420,9 +421,10 @@ class VersionsLockPluginIntegrationSpec extends IntegrationSpec {
         new File(subdir, 'build.gradle') << """
             apply plugin: 'java-library'
         """.stripIndent()
+        debug = true
```
Comment (ferozco) [inline src/test/groovy/com/palantir/gradle/versions/VersionsLockPluginIntegrationSpec.groovy]: nit: ensure to remove before merge

File: src/test/groovy/com/palantir/gradle/versions/VersionsLockPluginIntegrationSpec.groovy [hunk 7]
```diff
@@ -420,9 +421,10 @@ class VersionsLockPluginIntegrationSpec extends IntegrationSpec {
...
 
         expect:
-        runTasks('--write-locks')
+        runTasks('--write-locks', '-S')
```
Comment (ferozco) [inline src/test/groovy/com/palantir/gradle/versions/VersionsLockPluginIntegrationSpec.groovy]: samesies

**Review by ferozco — COMMENTED** [review]
looking pretty hot!

File: src/main/java/com/palantir/gradle/versions/VersionsLockPlugin.java (hunk 4 above)
Comment (dansanduleac) [reply src/main/java/com/palantir/gradle/versions/VersionsLockPlugin.java:426]: so turns out it's not a good idea to mark everything ORIGINAL as this then gets published in the module metadata file and causes potential issues for consumers in the future. Unless we only put that attribute onto other configurations that are not `runtimeElements` / `apiElements`, but then what's the point

File: src/main/java/com/palantir/gradle/versions/VersionsLockPlugin.java (hunk 1 above)
Comment (dansanduleac) [reply src/main/java/com/palantir/gradle/versions/VersionsLockPlugin.java:254]: Will do this as a FLUP

**Comment by iamdanfox** [comment]
10/10 PR desc mate.

**Review by iamdanfox — APPROVED** [review]

**Pull request merged**